api/agent/agent_launcher.go

```diff
@@ -0,0 +1,150 @@
+package agent
+
+import (
+	"context"
+	"time"
+
+	"github.com/fnproject/fn/api/models"
+)
+
+type WaitResult struct {
+	Slot Slot
+	Err  error
+}
+
+// tryGetToken attempts to fetch/acquire a token from slot queue without blocking.
+func tryGetToken(ch chan WaitResult) (Slot, error) {
+	select {
+	case s := <-ch:
+		return s.Slot, s.Err
+	default:
+		return nil, nil
+	}
+}
+
+// waitGetToken blocks and waits on both wait and token channels.
+func waitGetToken(ch chan WaitResult, wait chan struct{}) (Slot, error) {
+	select {
+	case s := <-ch:
+		return s.Slot, s.Err
+	case <-wait:
+	}
+	return nil, nil
+}
+
+// listenWaitResult listens [context, agent shutdown, slot channel] and pushes the result to one channel
+func (a *agent) listenWaitResult(ctx context.Context, call *call, slotChan chan *slotToken) chan WaitResult {
+	output := make(chan WaitResult)
+	dampen := make(chan struct{}, 1)
+
+	go func() {
+		dampen <- struct{}{}
+		for {
+			select {
+			case s := <-slotChan:
+				if call.slots.acquireSlot(s) {
+					if s.slot.Error() != nil {
+						s.slot.Close()
+						output <- WaitResult{nil, s.slot.Error()}
+						return
+					}
+					select {
```

Comment:
here select is needed, because we don't want to leak slot. But for other cases, we pre-close it already in this function so we only need this extra select here.